local-cli/runAndroid/runAndroid.js

```diff
@@ -78,6 +78,117 @@ function tryRunAdbReverse(device) {
+}
+
+function runOnSpecificDevice(args) {
+  let devices = adb.getDevices();
+  let selectedDevice = findMatchingDevice(devices, args.deviceIdFromList);
+  if (selectedDevice){
+    buildApk();
+    installAndLaunchOnDevice(args, selectedDevice);
+  } else {
+      if (devices){
+        console.log('Could not find device with the id: "' + args.deviceIdFromList + '".');
+        console.log('Choose one of the following:');
+        child_process.execFileSync('adb', ['devices'], {
+          stdio: [process.stdin, process.stdout, process.stderr],
+        });
+      } else {
+        console.log('No android devices connected.');
+      }
+  }
+}
+
+function buildApk() {
+  try {
+    const cmd = process.platform.startsWith('win')
+      ? 'gradlew.bat'
+      : './gradlew';
+
+    console.log(chalk.bold(
+      'Building the app...'
+    ));
+
+    child_process.execFileSync(cmd, ['build', '-x', 'lint'], {
+      stdio: [process.stdin, process.stdout, process.stderr],
+    });
+  } catch (e) {
+    console.log(chalk.red(
+      'Could not build the app on the device, read the error above for details.\n'
+    ));
+    // stderr is automatically piped from the gradle process, so the user
+    // should see the error already, there is no need to do
+    // `console.log(e.stderr)`
+    return Promise.reject();
+  }
+}
+
+function tryInstallAppOnDevice(args, device) {
+  try {
+    const pathToApk = 'app/build/outputs/apk/app-debug.apk';
+    const adbPath = getAdbPath();
+    const adbArgs = ['-s', device, 'install', pathToApk];
+    console.log(chalk.bold(
+      `Installing the app on the device (cd android && adb -s ${device} install ${pathToApk}`
+    ));
+    child_process.execFileSync(adbPath, adbArgs, {
+      stdio: [process.stdin, process.stdout, process.stderr],
+    });
+  } catch (e) {
+    console.log(e.message);
+    console.log(chalk.red(
+      'Could not install the app on the device, read the error above for details.\n'
+    ));
+    // stderr is automatically piped from the gradle process, so the user
+    // should see the error already, there is no need to do
+    // `console.log(e.stderr)`
+    return Promise.reject();
+  }
+}
+
+function tryLaunchAppOnDevice(device) {
+  try {
+    const packageName = fs.readFileSync(
+      'app/src/main/AndroidManifest.xml',
+      'utf8'
+    ).match(/package="(.+?)"/)[1];
+    const adbPath = getAdbPath();
+    const adbArgs = ['-s', device, 'shell', 'am', 'start', '-n', packageName + '/.MainActivity'];
+    console.log(chalk.bold(
+      `Starting the app on ${device} (${adbPath} ${adbArgs.join(' ')})...`
+    ));
+    child_process.spawnSync(adbPath, adbArgs, {stdio: 'inherit'});
+  } catch (e) {
+    console.log(chalk.red(
+      'adb invocation failed. Do you have adb in your PATH?'
+    ));
+    // stderr is automatically piped from the gradle process, so the user
+    // should see the error already, there is no need to do
+    // `console.log(e.stderr)`
+    return Promise.reject();
```

Comment:
Hi, thanks for the feedback.
i called it '--deviceIdFromList' because its not the actually device id. It is just the id created by adb. -> https://developer.android.com/studio/command-line/adb.html#commandsummary
However '--deviceId' might still be better, since it is easier to use.

I will do some refactoring and remove the promise on buildApk.